server/app/services/scheduler/filter/memory.rb

```diff
@@ -30,20 +33,49 @@ def for_service(service, instance_number, nodes)
         candidates
       end
 
+      # @param service [GridService]
+      # @return [Integer]
+      def resolve_memory_from_stats(service)
+        cache_key = "service_memory_peak:#{service.id}"
+        unless cache[cache_key] # aggregate needs to be cached manually
+          max_memory_usage = ContainerStat.where(
+            :grid_service_id => service.id,
+            :created_at.gt => 1.hour.ago
+          ).max(:'memory.usage')
+          if max_memory_usage
+            cache[cache_key] = max_memory_usage * 1.25
+          else
+            cache[cache_key] = 0.0
+          end
+        end
+        cache[cache_key].to_i
+      end
+
+      # @param service [GridService]
+      # @param instance_number [Integer]
+      # @param candidate [HostNode]
+      # @return [GridServiceInstance, NilClass]
+      def fetch_service_instance(service, instance_number, candidate = nil)
+        service.grid_service_instances.to_a.find { |i|
+          i.instance_number == instance_number && (candidate.nil? || i.host_node_id == candidate.id)
+        }
+      end
+
       # @param [HostNode] candidate
       # @param [Float] memory
       # @param [GridService] service
       # @param [Integer] instance_number
       def reject_candidate?(candidate, memory, service, instance_number)
-        return false if candidate.containers.service_instance(service, instance_number).first
+        return false if fetch_service_instance(service, instance_number, candidate)
         return true if candidate.mem_total.to_i < memory
 
-        node_stat = candidate.host_node_stats.latest
-        return false if node_stat.nil?
+        node_stat = candidate.latest_stats
+        return false if node_stat.empty?
```

Comment:
A little wary of this give false-positives during the first minute immediately after the upgrade, when the nodes haven't yet sent their latest stats, and the memory filter will return all available nodes. Could that lead to some unexpected turbulence during the upgrade? OTOH I suppose that is still acceptable...